src/backends/raster/bitmap_backend.f90

```diff
@@ -0,0 +1,166 @@
+            end do
+        end do
+
+        close(unit_num)
+    end subroutine save_to_ppm
+
+    subroutine fig_draw_pixel_i(canva, x, y, color)
+        !! very temporary
+        type(bitmap_canvas), intent(inout) :: canva
+        integer, intent(in) :: x, y
+        integer(pixel), intent(in) :: color
+    
+        if (x >= 0 .and. x < int(canva%width) .and. y >= 0 .and. y < int(canva%height)) then
+            canva%pixels(x, y) = color
+        end if
+    end subroutine fig_draw_pixel_i
+
+
+    subroutine write_circle(circ,canva)
+        !! TODO will implement fill later when i do lines
+        type(circle), intent(in) :: circ
+        class(bitmap_canvas), intent(inout) :: canva
+        integer(pixel) :: color 
+        integer :: x, y, d
+
+        color = rgb_to_int(circ%fill_color)
+        x = 0
+        y = int(circ%r)
+        d = 1 - int(circ%r)
+
+        do while (x < y)
+            if (d < 0) then
+                d = d + 2 * x + 3
+            else
+                d = d + 2 * (x - y) + 5
+                y = y - 1
+            end if
+            x = x + 1
+
+            call fig_draw_pixel_i(canva, int(circ%cx) + x, int(circ%cy) + y, color)
+            call fig_draw_pixel_i(canva, int(circ%cx) - x, int(circ%cy) + y, color)
+            call fig_draw_pixel_i(canva, int(circ%cx) + x, int(circ%cy) - y, color)
+            call fig_draw_pixel_i(canva, int(circ%cx) - x, int(circ%cy) - y, color)
+            call fig_draw_pixel_i(canva, int(circ%cx) + y, int(circ%cy) + x, color)
+            call fig_draw_pixel_i(canva, int(circ%cx) - y, int(circ%cy) + x, color)
+            call fig_draw_pixel_i(canva, int(circ%cx) + y, int(circ%cy) - x, color)
+            call fig_draw_pixel_i(canva, int(circ%cx) - y, int(circ%cy) - x, color)
+        end do
+
+        call fig_draw_pixel_i(canva, int(circ%cx) , int(circ%cy - circ%r), color)
+        call fig_draw_pixel_i(canva, int(circ%cx) , int(circ%cy + circ%r), color)
+        call fig_draw_pixel_i(canva, int(circ%cx - circ%r), int(circ%cy) , color)
+        call fig_draw_pixel_i(canva, int(circ%cx + circ%r), int(circ%cy) , color)
+    end subroutine write_circle
+
+    subroutine write_rectangle(rect,canva)
+        type(rectangle), intent(in) :: rect
+        class(bitmap_canvas), intent(inout) :: canva
+
+        integer(pixel) :: color
+        integer :: x, y
+        integer :: x_start, y_start
+        integer :: x_end, y_end
+        
+        color = rgb_to_int(rect%fill_color)
+        
+        x_start = max(int(rect%x),0)
+        y_start = max(int(rect%y),0)
+        x_end = int(min(rect%x + rect%width, canva%width))-1
+        y_end = int(min(rect%y + rect%height, canva%height))-1
+        
+        do y = y_start, y_end 
+            do x = x_start, x_end 
+                canva%pixels(x, y) = color
+            end do
+        end do
+
+    end subroutine write_rectangle
+    
+
+    subroutine write_shape(sh,canva)
```

Comment:
Sorry, I have just seen your comment.

I was intending to do something similar, but I couldn't find a good design to make `write_rectangle` work differently for SVG and bitmap without making them separate functions or creating two different types that inherit from an abstract rectangle.

Reply:
`rectangle` is a primitive, and each `canvas` type must have its own implementation of how to draw primitives. 
One way could be via a single polymorphic subroutine: 

```fortran
subroutine bmp_write_shape(c,s)
   class(bmp_canvas), intent(inout) :: c
   class(shape), intent(in) :: s

   ! Resolve shape
   select type (this_shape => s)
      type is (rectangle)
          call write_bmp_rectangle(c,this_shape)
      type is (square)
          call write_bmp_square(c,this_shape)
      class default
          stop 'shape class not supported by BMP canvas'
   end select 
```

2) Write the actual implementations as separate subroutines, i.e.

```fortran
subroutine write_bmp_rectangle(c,rect)
   class(bmp_canvas), intent(inout) :: c
   type(rectangle), intent(in) :: rect ! note non-polymorphic input
   ....
end subroutine write_bmp_rectangle
```

3) To force each canvas to have such drawing subroutine, you have to define it in the abstract class, which means you need an interface for it: 

```fortran
abstract interface
   subroutine canvas_draw_shape(c,s)
   class(TCanvas), intent(inout) :: c
   class(shape), intent(in) :: s
   end subroutine canvas_draw_shape
end interface
```

4) add the interface to the abstract class as `deferred`: this will force all canvases to have such an implementation defined

```fortran
type, abstract :: TCanvas
   contains
   procedure(canvas_draw_shape), deferred :: draw_shape
   procedure(canvas_draw_text), deferred :: draw_text
   generic :: draw => draw_shape, draw_text
end type TCanvas
```